Diviner flavor (#5553)
mlflow/mlflow: docs/source/models.rst

    a non-pyfunc artifact. The output of the native ``ARIMA.predict()`` when returning confidence intervals is not
    a recognized signature type.

Diviner (``diviner``) (Experimental)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
The ``diviner`` model flavor enables logging of

> `diviner models <https://databricks-diviner.readthedocs.io/en/latest/index.html>`_ in MLflow format via the

Author: This link is broken until diviner's repo moves to 'public status'. Any day now...

:py:func:`mlflow.diviner.save_model()` and :py:func:`mlflow.diviner.log_model()` methods. These methods also add the
``python_function`` flavor to the MLflow Models that they produce, allowing the model to be interpreted as generic
Python functions for inference via :py:func:`mlflow.pyfunc.load_model()`.
This loaded PyFunc model can only be scored with a DataFrame input.
You can also use the :py:func:`mlflow.diviner.load_model()` method to load MLflow Models with the ``diviner``
model flavor in native diviner formats.

Diviner Types
~~~~~~~~~~~~~
Diviner is a library that provides an orchestration framework for performing time series forecasting on groups of
related series. Forecasting in ``diviner`` is accomplished through wrapping popular open source libraries such as
`prophet <https://facebook.github.io/prophet/>`_ and `pmdarima <http://alkaline-ml.com/pmdarima/>`_. The ``diviner``
library offers a simplified set of APIs to generate many time series forecasts from a single input DataFrame and a

> unified high-level API.

Author: great edit!

metrics table.

To eliminate this issue for large-scale forecasting, the metrics and parameters for ``diviner`` are extracted as a
grouping key indexed ``Pandas DataFrame``, as shown below for example (float values truncated for visibility):

===================== ======= ========== ========== ====== ====== ==== ===== =====
grouping_key_columns  country city       mse        rmse   mae    mape mdape smape
===================== ======= ========== ========== ====== ====== ==== ===== =====
"('country', 'city')" CA      Toronto    8276851.6  2801.7 2417.7 0.16 0.16  0.159
"('country', 'city')" MX      MexicoCity 3548872.4  1833.8 1584.5 0.15 0.16  0.159
"('country', 'city')" US      NewYork    3167846.4  1732.4 1498.2 0.15 0.16  0.158
"('country', 'city')" US      Boston     14082666.4 3653.2 3156.2 0.15 0.16  0.159
===================== ======= ========== ========== ====== ====== ==== ===== =====

There are two recommended means of logging the metrics and parameters from a ``diviner`` model :

* Writing the DataFrames to local storage and using :py:func:`mlflow.log_artifacts`

.. code-block:: py

    import os
    import mlflow

>     local = "/tmp/forecast/diviner"
>     os.makedirs(local, exist_ok=True)
>     params = model.extract_model_params()
>     metrics = model.cross_validate_and_score(
>         horizon="72 hours",
>         period="240 hours",
>         initial="480 hours",
>         parallel="threads",
>         rolling_window=0.1,
>         monthly=False,
>     )
>     params.to_csv(f"{local}/params.csv", index=False, header=True)
>     metrics.to_csv(f"{local}/metrics.csv", index=False, header=True)
> 
>     mlflow.log_artifacts(local, artifact_path="data")

Reviewer: ```suggestion
    with tempfile.TemporaryDirectory() as tmpdir:
        params = model.extract_model_params()
        metrics = model.cross_validate_and_score(
            horizon="72 hours",
            period="240 hours",
            initial="480 hours",
            parallel="threads",
            rolling_window=0.1,
            monthly=False,
        )
        params.to_csv(f"{tmpdir}/params.csv", index=False, header=True)
        metrics.to_csv(f"{tmpdir}/metrics.csv", index=False, header=True)

        mlflow.log_artifacts(local, artifact_path="data")
```

Can we use a temporary directory?

Author: absolutely fantastic idea! Changing.

* Writing directly as a JSON artifact using :py:func:`mlflow.log_dict`

.. note::
    The parameters extract from ``diviner`` models *may require* casting (or dropping of columns) if using the
    ``pd.DataFrame.to_dict()`` approach due to the inability of this method to serialize objects.

.. code-block:: py

    import mlflow

    params = model.extract_model_params()
    metrics = model.cross_validate_and_score(
        horizon="72 hours",
        period="240 hours",
        initial="480 hours",
        parallel="threads",
        rolling_window=0.1,
        monthly=False,
    )
    params["t_scale"] = params["t_scale"].astype(str)
    params["start"] = params["start"].astype(str)
    params = params.drop("stan_backend", axis=1)

    mlflow.log_dict(params.to_dict(), "params.json")
    mlflow.log_dict(metrics.to_dict(), "metrics.json")

Logging of the model artifact is shown in the ``pyfunc`` example below.

Diviner pyfunc usage
~~~~~~~~~~~~~~~~~~~~
The ``diviner`` interface for the ``pyfunc`` type is a patched API that is defined by the configuration supplied

> in the argument to ``predict()``. This configuration is supplied as a single row ``Pandas DataFrame``.

Author: much better

As this configuration is dependent upon the underlying model type (i.e., the ``diviner.GroupedProphet.forecast()`` method has
a different signature than does ``diviner.GroupedPmdarima.predict()``), resolution of provided argument names will be

> optimistically coerced into the appropriate argument keys.

Author: nice change

.. note::

>     Diviner models support both "full group" and "partial group" forecasting. If a column named "groups" is present

Author: good catch
